sds_data_manager/lambda_images/instruments/batch_starter.py

```diff
@@ -0,0 +1,425 @@
+        for d in dependencies
+    ]
+
+    # Check if the items in dependencies are all present in result_list
+    if set(tuple(item.items()) for item in dependencies).issubset(
+        set(tuple(item.items()) for item in result_list)
+    ):
+        print("All dependencies are found.")
+        return True
+    else:
+        print("Some dependencies are missing.")
+        return False
+
+
+def prepare_data(instruments_to_process):
+    """
+    Groups input data by 'instrument' and 'level', and aggregates the dates
+    for each group into a list.
+
+    Parameters
+    ----------
+    instruments_to_process : list of dict
+        A list of dictionaries which is not aggregated.
+
+    Returns
+    -------
+    grouped_dict: dict
+        A dictionary of instruments, each containing a dictionary of
+        levels with a list of dates.
+    """
+    grouped_data = defaultdict(lambda: defaultdict(list))
+    for item in instruments_to_process:
+        instrument = item["instrument"]
+        level = item["level"]
+        date = item["date"]
+        grouped_data[instrument][level].append(date)
+
+    grouped_dict = {
+        instrument: dict(levels) for instrument, levels in grouped_data.items()
+    }
+
+    return grouped_dict
+
+
+def lambda_handler(event: dict, context):
+    """Handler function"""
+    logger.info(f"Event: {event}")
+    logger.info(f"Context: {context}")
+
+    instrument = os.environ.get("INSTRUMENT")
+    instrument_dependents = os.environ.get("INSTRUMENT_DEPENDENTS")
+    state_machine_arn = os.environ.get("STATE_MACHINE_ARN")
+    db_secret_arn = os.environ.get("SECRET_ARN")
+
+    filename = get_filename_from_event(event)
+
+    with db_connect(db_secret_arn) as conn:
+        with conn.cursor() as cur:
+            # get details of the object
+            level, version, process_dates = get_process_details(
+                cur, instrument, filename
+            )
+            # query dependents to see if they have been ingested
+            ingested_dependents = query_dependents(
+                cur, version, process_dates, instrument_dependents[level]
+            )
+            # remove dependents that have been ingested
+            uningested = remove_ingested(
+                ingested_dependents, instrument_dependents[level], process_dates
+            )
+
+            # Check if uningested is empty
+            if not uningested:
+                logger.info(
+                    "No uningested dependents found. Skipping further processing."
+                )
+                return
+
+            # decide if we have sufficient dependencies
+            # for each dependent to process
+            instruments_to_process = query_dependencies(cur, uningested, version)
+
+            # No instruments to process
+            if not instruments_to_process:
+                logger.info("No instruments_to_process. Skipping further processing.")
+                return
```

Comment:
If you can add some example in the comments of what it will look like at each of these operation, it will help so much to follow and understand. I know it's a complicated logic and it may help us even when we come back to look at this logic again after some time.